tangos/input_handlers/ramsesHOP.py

```diff
@@ -0,0 +1,137 @@
+                      object_typetag="halo", output_handler_for_ts2=None):
+
+        f1 = self.load_timestep(ts1).dm
+        h1 = self._construct_halo_cat(ts1, object_typetag)
+
+        if output_handler_for_ts2 is None:
+            f2 = self.load_timestep(ts2).dm
+            h2 = self._construct_halo_cat(ts2, object_typetag)
+        else:
+            f2 = output_handler_for_ts2.load_timestep(ts2).dm
+            h2 = output_handler_for_ts2._construct_halo_cat(ts2, object_typetag)
+
+        bridge = pynbody.bridge.OrderBridge(f1, f2, monotonic=False)
+
+        return bridge.fuzzy_match_catalog(halo_min, halo_max, threshold=threshold, only_family=pynbody.family.dm,
+                                          groups_1=h1, groups_2=h2)
+
+
+
+class RamsesAdaptaHOPInputHandler(RamsesHOPInputHandler):
+    """ Handling Ramses outputs with AdaptaHOP halo and subhalo finding """
+
+    patterns = ["output_0????"]
+    auxiliary_file_patterns = ["tree_bricks???"]
+
+    def _exclude_adaptahop_precalculated_properties(self):
+        return ["members", "timestep", "level", "host_id", "first_subhalo_id", "next_subhalo_id",
+                "x", "y", "z", "vx", "vy", "vz", "lx", "ly", "lz",
+                "contaminated", "m_contam", "mtot_contam", "n_contam", "ntot_contam"]
+
+    def _include_additional_properties_derived_from_adaptahop(self):
+        # I don't know what actually are vx, vy, vz and lx, ly, lz, but they could be added as reformatted 3D arrays here
+        return ["parent", "child", "shrink_center", "contamination_fraction"]
+
+    @staticmethod
+    def _reformat_center(adaptahop_halo):
+        return np.array([adaptahop_halo.properties['x'], adaptahop_halo.properties['y'], adaptahop_halo.properties['z']])
+
+    @staticmethod
+    def _compute_contamination_fraction(adaptahop_halo):
+        return float(adaptahop_halo.properties['ntot_contam'] / adaptahop_halo.properties['ntot'])
+
+    def available_object_property_names_for_timestep(self, ts_extension, object_typetag):
+        h = self._construct_halo_cat(ts_extension, object_typetag)
+
+        halo_attributes = list(h._halo_attributes)
+        if h._read_contamination:
+            halo_attributes.extend(h._halo_attributes_contam)
+
+        attrs = chain.from_iterable(tuple(always_iterable(attr)) for (attr, _len, _dtype) in halo_attributes)
+
+        # Import all precalculated properties except conflicting ones with Tangos syntax
+        property_list = [attr for attr in attrs if attr not in self._exclude_adaptahop_precalculated_properties()]
+        # Add additional properties that are baisc to most Tangos databases derived
+        property_list += self._include_additional_properties_derived_from_adaptahop()
+        return property_list
+
+    @staticmethod
+    def _resolve_units(value):
+        import pynbody
+        # TODO Solve the fact that AdapataHOP stores distances in Mpc, when Tangos physical units are kpc
```

Comment:
One thing to double-check is that AdaptaHOP may assume that `1Mpc = 3.08 10^24 cm` **exactly** (at least it used to, and I haven't checked it still does) instead of `3.08567758 10^24 cm`.
This may be the source of some trouble when dealing with the centre of halos which may be off by a bit, especially when using deep zooms.

Reply:
Yes you're right, I just checked AdaptaHOP assumes 3.08e24.

I think this knowledge would be better dealt at the loader level, i.e. patching pynbody to (i) provide the exact correct number we expect from the halo finder, and (ii) in an array rather than a unit so that it can be converted easily to whichever other unit we wish ?

Reply:
I had forgotten, but pynbody's AdaptaHOP loader already does the right thing, see 
https://github.com/pynbody/pynbody/blob/23913471a6d6511a024a5fde83b9fcd3aef7b2ed/pynbody/halo/adaptahop.py#L246-L251.